mm2src/coins/qrc20.rs

```diff
@@ -0,0 +1,1059 @@
+                  GenerateTransactionError, UtxoAddressFormat, UtxoCoinFields, UtxoCommonOps, UtxoTx,
+                  VerboseTransactionFrom, UTXO_LOCK};
+use crate::{FoundSwapTxSpend, HistorySyncState, MarketCoinOps, MmCoin, SwapOps, TradeFee, TransactionDetails,
+            TransactionEnum, TransactionFut, ValidateAddressResult, WithdrawFee, WithdrawRequest};
+use async_trait::async_trait;
+use bigdecimal::BigDecimal;
+use bitcrypto::{dhash160, sha256};
+use chain::TransactionOutput;
+use common::block_on;
+use common::executor::Timer;
+use common::jsonrpc_client::{JsonRpcClient, JsonRpcError, JsonRpcRequest, RpcRes};
+use common::mm_ctx::MmArc;
+use ethabi::{Function, Token};
+use ethereum_types::{H160, U256};
+use futures::compat::Future01CompatExt;
+use futures::{FutureExt, TryFutureExt};
+use futures01::Future;
+use gstuff::now_ms;
+use keys::bytes::Bytes as ScriptBytes;
+use keys::{Address as UtxoAddress, Public};
+#[cfg(test)] use mocktopus::macros::*;
+use rpc::v1::types::{Bytes as BytesJson, Transaction as RpcTransaction, H160 as H160Json, H256 as H256Json};
+use script::{Builder as ScriptBuilder, Opcode, Script, TransactionInputSigner};
+use script_pubkey::generate_contract_call_script_pubkey;
+use serde_json::{self as json, Value as Json};
+use serialization::deserialize;
+use serialization::serialize;
+use std::ops::{Deref, Neg};
+use std::str::FromStr;
+use std::sync::Arc;
+
+mod history;
+#[cfg(test)] mod qrc20_tests;
+pub mod rpc_electrum;
+mod script_pubkey;
+mod swap_ops;
+
+/// Qtum amount is always 0 for the QRC20 UTXO outputs,
+/// because we should pay only a fee in Qtum to send the QRC20 transaction.
+const OUTPUT_QTUM_AMOUNT: u64 = 0;
+const QRC20_GAS_LIMIT_DEFAULT: u64 = 100_000;
+const QRC20_GAS_PRICE_DEFAULT: u64 = 40;
+const QRC20_SWAP_GAS_REQUIRED: u64 = QRC20_GAS_LIMIT_DEFAULT * 3;
+const QRC20_DUST: u64 = 0;
+// Keccak-256 hash of `Transfer` event
+const QRC20_TRANSFER_TOPIC: &str = "ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef";
+#[allow(dead_code)]
+const QRC20_APPROVE_TOPIC: &str = "8c5be1e5ebec7d5bd14f71427d1e84f3dd0314c0f7b2291e5b200ac8c7c3b925";
+#[allow(dead_code)]
+const QRC20_PAYMENT_SENT_TOPIC: &str = "ccc9c05183599bd3135da606eaaf535daffe256e9de33c048014cffcccd4ad57";
+
+pub async fn qrc20_coin_from_conf_and_request(
+    ctx: &MmArc,
+    ticker: &str,
+    platform: &str,
+    conf: &Json,
+    req: &Json,
+    priv_key: &[u8],
+    contract_address: H160,
+) -> Result<Qrc20Coin, String> {
+    if let Some("enable") = req["method"].as_str() {
+        return ERR!("Native mode not supported yet for QRC20");
+    }
+    let swap_contract_address = match req["swap_contract_address"].as_str() {
+        Some(address) => try_s!(qrc20_addr_from_str(address)),
+        None => return ERR!("\"swap_contract_address\" field is expected"),
+    };
+    let utxo = try_s!(utxo_fields_from_conf_and_request(ctx, ticker, conf, req, priv_key, QRC20_DUST).await);
```

Comment:
If `decimals` are not set in request `utxo_fields_from_conf_and_request` will default to 8 which might be unsafe. We should request decimals directly from QRC20 smart contract and use them. And also return the error if decimals are not available similar to ERC20 initialization.

https://github.com/KomodoPlatform/atomicDEX-API/blob/31d1803b0a5b6439e51ed85ab41cea1accff8546/mm2src/coins/eth.rs#L2601-L2608

Reply:
Hmm, I guess it could lead to loss of funds. I will fix it right now

Reply:
Done

Reply:
It's a temporary solution, because Native RPC API doesn't have a method to get token decimals. I will change it with `decimals` like `balanceOf`